src/waitress/channel.py

```diff
@@ -360,88 +381,101 @@ def _flush_outbufs_below_high_watermark(self):
-                    # much data in the outbufs waiting to be flushed
-                    # NB: currently readable() returns False while we are
-                    # flushing data so we know no new requests will come in
-                    # that we need to account for, otherwise it'd be better
-                    # to do this check at the start of the request instead of
-                    # at the end to account for consecutive service() calls
-
-                    if len(self.requests) > 1:
-                        self._flush_outbufs_below_high_watermark()
-
-                    # this is a little hacky but basically it's forcing the
-                    # next request to create a new outbuf to avoid sharing
-                    # outbufs across requests which can cause outbufs to
-                    # not be deallocated regularly when a connection is open
-                    # for a long time
-
-                    if self.current_outbuf_count > 0:
-                        self.current_outbuf_count = self.adj.outbuf_high_watermark
-
-                    request = self.requests.pop(0)
+        if task.close_on_finish:
+            with self.task_lock:
+                self.close_when_flushed = True
+
+                for request in self.requests:
                     request.close()
+                self.requests = []
+        else:
+            # before processing a new request, ensure there is not too
+            # much data in the outbufs waiting to be flushed
+            # NB: currently readable() returns False while we are
+            # flushing data so we know no new requests will come in
+            # that we need to account for, otherwise it'd be better
+            # to do this check at the start of the request instead of
+            # at the end to account for consecutive service() calls
+
+            if len(self.requests) > 1:
+                self._flush_outbufs_below_high_watermark()
+
+            # this is a little hacky but basically it's forcing the
+            # next request to create a new outbuf to avoid sharing
+            # outbufs across requests which can cause outbufs to
+            # not be deallocated regularly when a connection is open
+            # for a long time
+
+            if self.current_outbuf_count > 0:
+                self.current_outbuf_count = self.adj.outbuf_high_watermark
+
+            request.close()
+
+            # Add new task to process the next request
+            with self.task_lock:
+                self.requests.pop(0)
+                if self.connected and self.requests:
+                    self.server.add_task(self)
+                elif (
+                    self.connected
+                    and self.request is not None
+                    and self.request.expect_continue
+                    and self.request.headers_finished
+                    and not self.sent_continue
+                ):
```

Comment:
`self.request` is shared between the asyncore thread and this thread but not protected by any particular lock. The `task_lock` is acquired here which is great, so It seems like some more effort needs to be made to do something inside `channel.received()` to protect the usage there.

Reply:
I guess you are right.
I used the `task_lock` for all operations on `self.requests` and now also `self.request`. My question is: should I rename it to `requests_lock`? The original purpose of the lock was to protect multiple calls to `service` from actually running, but I guess this changed now.

Reply:
We need to make sure that `channel.received()` does NOT block for extended periods of time. I haven't gone through the code with a fine-tooth comb yet, but please be aware that if we block in `channel.received()` we are blocking inside the main thread... how long and how often do we hold the locks? The original lock was held for the duration of request processing, if that is no longer the case then I think it may be fine.

I am all for renaming variables to match their intent, so I am a +1 on that!

Reply:
I talked about this with @bertjwregeer offline but I'm quite certain that the original `task_lock` is completely unnecessary in waitress right now because I don't think we ever scheduled the channel concurrently - which is what it was protecting against - ensuring that the channel was not handled by multiple worker threads. With that in mind renaming it totally makes sense.

It's a great point about the locking concern within received. In the previous version of the code received was never invoked while processing a request and now it may be. @viktordick has drastically reduced the time that the lock is held versus what's done in master such that it's not wrapping the entire task lifecycle anymore. There isn't much that really concerns me with how it's being handled right now - it's only really held in 3 places and two of them are around send_continue, with the final being the close_on_finish teardown.

The send_continue is potentially long running but I do not see how it's any worse than on master.